File: sdk/cpp/symbolchain/symbol/Network.py
import base64
import datetime
import hashlib
from binascii import unhexlify

from ..ByteArray import ByteArray
from ..CryptoTypes import Hash256
from ..Network import Network as BasicNetwork
from ..NetworkTimestamp import NetworkTimestamp as BasicNetworkTimestamp
from ..NetworkTimestamp import NetworkTimestampDatetimeConverter
from ..sc import NamespaceId


class NetworkTimestamp(BasicNetworkTimestamp):
	"""Represents a symbol network timestamp with millisecond resolution."""

	def add_milliseconds(self, count):
		"""Adds a specified number of milliseconds to this timestamp."""
		return NetworkTimestamp(self.timestamp + count)

	def add_seconds(self, count):
		return self.add_milliseconds(1000 * count)


class Address(ByteArray):
	"""Represents a Symbol address."""

	SIZE = 24
	ENCODED_SIZE = 39

	def __init__(self, address):
		"""Creates an address from a decoded or encoded address."""
		raw_bytes = address
		if isinstance(address, str):
			raw_bytes = base64.b32decode(address + 'A')[0:-1]
		elif isinstance(address, Address):
			raw_bytes = address.bytes

		super().__init__(self.SIZE, raw_bytes, Address)

	def to_namespace_id(self):
		"""Attempts to convert this address into a namespace id."""

		if not self.bytes[0] & 0x01:
			return None

		return NamespaceId(int.from_bytes(self.bytes[1:9], 'little'))

	def __str__(self):
		return base64.b32encode(self.bytes + bytes(0)).decode('utf8')[0:-1]

	def __repr__(self):
		return f'Address(\'{str(self)}\')'

	@staticmethod
	def from_decoded_address_hex_string(hex_string):
		"""Creates an address from a decoded address hex string (typically from REST)."""

		return Address(unhexlify(hex_string))

	@staticmethod
	def from_namespace_id(namespace_id, network_identifier):
		"""Creates an address from a namespace id."""

		return Address(bytes([network_identifier + 1]) + namespace_id.value.to_bytes(8, 'little') + bytes([0] * (Address.SIZE - 9)))


class Network(BasicNetwork):
	"""Represents a Symbol network."""

	def __init__(self, name, identifier, epoch_time, generation_hash_seed=None):
		"""Creates a new network with the specified properties."""
		super().__init__(name, identifier, NetworkTimestampDatetimeConverter(epoch_time, 'milliseconds'), Address, NetworkTimestamp)
		self.generation_hash_seed = generation_hash_seed

	def address_hasher(self):
		return hashlib.sha3_256()

	def create_address(self, address_without_checksum, checksum):
		return Address(address_without_checksum + checksum[0:3])


Network.MAINNET = Network(
	'mainnet',
	0x68,
	datetime.datetime(2021, 3, 16, 0, 6, 25, tzinfo=datetime.timezone.utc),
	Hash256('57F7DA205008026C776CB6AED843393F04CD458E0AA2D9F1D5F31A402072B2D6'))
Network.TESTNET = Network(
	'testnet',
	0x98,
	datetime.datetime(2022, 10, 31, 21, 7, 47, tzinfo=datetime.timezone.utc),
	Hash256('49D6E1CE276A85B70EAFE52349AACCA389302E7A9754BCF1221E79494FC665A4'))
Network.NETWORKS = [Network.MAINNET, Network.TESTNET]

File: sdk/java/src/main/java/org/symbol/sdk/symbol/Network.java
package org.symbol.sdk.symbol;

import java.time.Instant;
import java.util.Arrays;
import java.util.List;

import org.bouncycastle.crypto.digests.SHA3Digest;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.NetworkTimestamp.NetworkTimestampDatetimeConverter;
import org.symbol.sdk.NetworkTimestamp.NetworkTimestampDatetimeConverter.TimeUnit;
import org.symbol.sdk.utils.ArrayHelpers;

/**
 * Represents a Symbol network.
 */
public final class Network extends org.symbol.sdk.Network<Address, NetworkTimestamp> {
	/** Symbol main network. */
	public static final Network MAINNET = new Network("mainnet", (byte) 0x68, Instant.parse("2021-03-16T00:06:25Z"),
			new CryptoTypes.Hash256("57F7DA205008026C776CB6AED843393F04CD458E0AA2D9F1D5F31A402072B2D6"));

	/** Symbol test network. */
	public static final Network TESTNET = new Network("testnet", (byte) 0x98, Instant.parse("2022-10-31T21:07:47Z"),
			new CryptoTypes.Hash256("49D6E1CE276A85B70EAFE52349AACCA389302E7A9754BCF1221E79494FC665A4"));

	/** Symbol well-known networks. */
	public static final List<Network> NETWORKS = List.of(MAINNET, TESTNET);

	/** Network generation hash seed. */
	public final CryptoTypes.Hash256 generationHashSeed;

	/**
	 * Creates a new Symbol network.
	 *
	 * @param name Network name.
	 * @param identifier Network identifier byte.
	 * @param epochTime Network epoch time.
	 * @param generationHashSeed Network generation hash seed.
	 */
	public Network(final String name, final byte identifier, final Instant epochTime, final CryptoTypes.Hash256 generationHashSeed) {
		super(name, identifier, new NetworkTimestampDatetimeConverter(epochTime, TimeUnit.MILLISECONDS), () -> new SHA3Digest(256),
				// Symbol addresses use a 3-byte checksum (the base derives a 4-byte one, so truncate)
				(addressWithoutChecksum, checksum) -> new Address(ArrayHelpers.concat(addressWithoutChecksum, Arrays.copyOf(checksum, 3))),
				Address.ENCODED_SIZE, Address::new, timestamp -> new NetworkTimestamp((long) timestamp));
		this.generationHashSeed = generationHashSeed;
	}
}
